File: Stack/C++/Implementation/Linked-List/Stack.cpp
//
// Created by vivek on 01-01-2023.
//

#include <iostream>
#include "Stack.h"

Stack::Stack() : top(nullptr) {}            // Constructor

void Stack::push(int x) {           // Push operation: insert element into stack
    if (empty()) {
        top = new Node(x);
        return;
    }
    Node *t = new Node(x);
    t->next = top;
    top = t;
}

int Stack::pop() {         // Pop operation: remove element from stack
    if (this->empty()) {
        std::cerr << "Stack underflow!\n";
        return -1;
    }

    int x = top->data;
    Node *t = top;
    top = top->next ? top->next : nullptr;
    delete t;
    return x;
}

std::ostream &operator<<(std::ostream &os, const Stack &stack) {            // Printing the stack
    Node *t = stack.top;
    while (t) {
        os << t->data << std::endl;
        t = t->next;
    }
    os << std::endl;
    return os;
}

int Stack::peek() {         // Getting the top element of the stack
    return empty() ? -1 : this->top->data;
}

Stack::~Stack() {           // Destructor to free up heap memory used by linked list
    Node *t = this->top;
    while (t) {
        top = top->next;
        delete t;
        t = top;
    }
}

File: Stack/Java/Stack/src/implementation/linkedlist/Stack.java
package implementation.linkedlist;

public class Stack {
    private Node top;

    public Stack() {
        top = null;
    }

    public void push(int x) {
        if (isEmpty()) {
            top = new Node(x);
        } else {
            Node t = new Node(x);
            t.next = top;
            top = t;
        }
    }

    public int pop() {
        if (isEmpty()) {
            System.err.println("Stack underflow!");
            return -1;
        }

        int x = top.data;
        top = (top.next != null) ? top.next : null;
        return x;
    }

    public int peek() {
        return isEmpty() ? -1 : top.data;
    }

    public boolean isEmpty() {
        return top == null;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        Node t = top;
        while (t != null) {
            sb.append(t.data).append("\n");
            t = t.next;
        }
        sb.append("\n");
        return sb.toString();
    }

    private class Node {
        private final int data;
        private Node next;

        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }
}

